File: Algorithms/src/hard/MystiXORama.java
import java.io.*;
import java.util.*;

public class Main {

    // ---------- Fast Input ----------
    static class FastScanner {
        private final byte[] data;
        private int idx = 0;

        FastScanner() throws IOException {
            data = System.in.readAllBytes();
        }

        long nextLong() {
            long num = 0;
            int sign = 1;
            while (idx < data.length && (data[idx] == ' ' || data[idx] == '\n' || data[idx] == '\r')) idx++;
            if (data[idx] == '-') {
                sign = -1;
                idx++;
            }
            while (idx < data.length && data[idx] >= '0') {
                num = num * 10 + (data[idx] - '0');
                idx++;
            }
            return num * sign;
        }

        int nextInt() {
            return (int) nextLong();
        }
    }

    // ---------- Segment Tree ----------
    static class SegTree {
        int n;
        long[] t;

        SegTree(int n) {
            this.n = n;
            t = new long[2 * n + 5];
        }

        void modify(int p, long value) {
            p += n;
            t[p] = value;
            while (p > 1) {
                p >>= 1;
                t[p] = t[p << 1] + t[p << 1 | 1];
            }
        }

        long query(int l, int r) {
            r++;
            long res = 0;
            for (l += n, r += n; l < r; l >>= 1, r >>= 1) {
                if ((l & 1) == 1) res += t[l++];
                if ((r & 1) == 1) res += t[--r];
            }
            return res;
        }
    }

    // ---------- Query ----------
    static class Query {
        int l, r, x, y, ind;
    }

    static int S;

    static boolean cmp(Query A, Query B) {
        if (A.l / S != B.l / S) return A.l < B.l;
        if ((A.l / S) % 2 == 1) return A.r > B.r;
        return A.r < B.r;
    }

    // ---------- Solve ----------
    static void solve(FastScanner fs) {
        int n = fs.nextInt();

        long[] a = new long[n + 1];
        for (int i = 1; i <= n; i++) {
            a[i] = fs.nextLong();
        }

        SegTree st = new SegTree(n + 5);
        S = (int) Math.sqrt(n);

        int q = fs.nextInt();
        Query[] Q = new Query[q];
        for (int i = 0; i < q; i++) {
            Q[i] = new Query();
            Q[i].l = fs.nextInt();
            Q[i].r = fs.nextInt();
            Q[i].x = fs.nextInt();
            Q[i].y = fs.nextInt();
            Q[i].ind = i;
        }

        Arrays.sort(Q, (A, B) -> {
    if (A.l / S != B.l / S)
        return Integer.compare(A.l, B.l);

    if ((A.l / S) % 2 == 1)
        return Integer.compare(B.r, A.r);

    return Integer.compare(A.r, B.r);
    });


        long[] ans = new long[q];
        long[] arr = new long[n + 1];
        HashMap<Long, Integer> freq = new HashMap<>();

        // update function (same logic as lambda in C++)
        class Updater {
            void apply(int idx, int delta) {
                long val = a[idx];
                int f = freq.getOrDefault(val, 0);

                arr[f] ^= val;
                st.modify(f, arr[f]);

                f += delta;
                freq.put(val, f);

                arr[f] ^= val;
                st.modify(f, arr[f]);
            }
        }

        Updater updater = new Updater();

        int curL = Q[0].l;
        int curR = Q[0].l - 1;

        for (int i = 0; i < q; i++) {
            while (curL > Q[i].l) updater.apply(--curL, 1);
            while (curR < Q[i].r) updater.apply(++curR, 1);
            while (curL < Q[i].l) updater.apply(curL++, -1);
            while (curR > Q[i].r) updater.apply(curR--, -1);

            ans[Q[i].ind] = st.query(Q[i].x, Q[i].y);
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < q; i++) {
            sb.append(ans[i]).append('\n');
        }
        System.out.print(sb);
    }

    // ---------- Main ----------
    public static void main(String[] args) throws Exception {
        FastScanner fs = new FastScanner();
        int T = fs.nextInt();
        while (T-- > 0) {
            solve(fs);
        }
    }
}

File: Algorithms/src/hard/MystiXORama.cpp
#include <bits/stdc++.h>
using namespace std;
#define int long long int


struct SegTree {
    int n;
    vector<int> t;
    SegTree(int n_) {
        n = n_;
        t.resize(2 * n + 5, 0);
    }
     
    void modify(int p, int value) {
      for (t[p += n] = value; p > 1; p >>= 1) t[p>>1] = t[p] + t[p^1];
    }
     
    int query(int l, int r) {
      r++; 
      int res = 0;
      for (l += n, r += n; l < r; l >>= 1, r >>= 1) {
        if (l&1) res += t[l++];
        if (r&1) res += t[--r];
      }
      return res;
    }
};

struct Query {
    int l, r, x, y, ind;
};

int S;

bool cmp(Query A, Query B){
  if (A.l / S != B.l / S) return A.l < B.l;
  return A.l / S % 2 ? A.r > B.r : A.r < B.r;
}

void solve() {

    int n;
    cin >> n;

    vector<int> a(n + 1);
    for(int i = 1; i <= n; i++) {
        cin >> a[i];
    }

    SegTree st(n + 5);

    S = sqrt(n);

    int q;
    cin >> q;

    vector<int> ans(q), arr(n + 1);
    vector<Query> Q(q);
    map<int, int> freq;
    for(int i = 0; i < q; i++) {
        int l, r, x, y;
        cin >> Q[i].l >> Q[i].r >> Q[i].x >> Q[i].y;
        Q[i].ind = i;
    }

    sort(Q.begin(), Q.end(), cmp);

    auto check = [&](int ind, int f) {
        arr[freq[a[ind]]] ^= a[ind];
        st.modify(freq[a[ind]], arr[freq[a[ind]]]);
        freq[a[ind]] += f;
        arr[freq[a[ind]]] ^= a[ind];
        st.modify(freq[a[ind]], arr[freq[a[ind]]]);
    };

    int curL = Q[0].l, curR = Q[0].l - 1;
    for(int i = 0; i < q; i++) {
        while (curL > Q[i].l) check(--curL, 1);
        while (curR < Q[i].r) check(++curR, 1);
        while (curL < Q[i].l) check(curL++, -1);
        while (curR > Q[i].r) check(curR--, -1);
        ans[Q[i].ind] = st.query(Q[i].x, Q[i].y);
    }

    for(int i = 0; i < q; i++) {
        cout << ans[i] << endl;
    }
}


#undef int 
int main() {
#define int long long int
ios_base::sync_with_stdio(false); 
cin.tie(0); 
cout.tie(0);

    int T;
    cin >> T;

    for(int tc = 1; tc <= T; tc++){
        // cout << "Case #" << tc << ": ";
        solve();
    }

return 0;  

}
